From 141442d8dd6d38111d0e526e0ef3bc94355f47f8 Mon Sep 17 00:00:00 2001
From: Haithem Salah <haithem.salah@intel.com>
Date: Tue, 18 Nov 2014 16:51:21 +0100
Subject: [PATCH 160/164] atomisp2: force ISP power down to avoid punit
 timeout

If CSS is waiting for more frames from atomisp2 driver while
atomisp2 driver wants to stop ISP, the stopping fails since
CSS doesn't listen the stop request in that case. Then, when
atomisp2 driver requests punit to power down ISP, it causes
punit timeout, because punit waits for ISP to go idle which
never happens.

This patch sets SRSE register to 2b'11 before requesting
punit to power off ISP, which causes ISP to be powered down
unconditionally without waiting it to go idle.

Change-Id: I686a180602e74d8d3c21c7225036dffc4cb135a8
Tracked-On: https://jira01.devtools.intel.com/browse/IMINAN-11132
Signed-off-by: Haithem Salah <haithem.salah@intel.com>
Signed-off-by: Tuukka Toivonen <tuukka.toivonen@intel.com>
Reviewed-on: https://android.intel.com:443/304414
Reviewed-on: https://icggerrit.ir.intel.com/30866
Approver: Wei Tao <wei.tao@intel.com>
Build: Wei Tao <wei.tao@intel.com>
Reviewed-by: Wei Tao <wei.tao@intel.com>
Maintainer: Wei Tao <wei.tao@intel.com>
---
 .../pci/atomisp2/atomisp_driver/atomisp_ioctl.c    |    5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_ioctl.c b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_ioctl.c
index a7aa69b..5c13bfd 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_ioctl.c
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_ioctl.c
@@ -2021,6 +2021,11 @@ stopsensor:
 				atomisp_destroy_pipes_stream_force(&isp->
 						asd[i]);
 		}
+
+		/* disable  PUNIT/ISP acknowlede/handshake - SRSE=3 */
+		pci_write_config_dword(isp->pdev, PCI_I_CONTROL, isp->saved_regs.i_control |
+				MRFLD_PCI_I_CONTROL_SRSE_RESET_MASK);
+		dev_err(isp->dev, "atomisp_reset");
 		atomisp_reset(isp);
 		isp->isp_timeout = false;
 	}
-- 
1.7.9.5

